apps/gameserver/src/Party.cpp

```
@@ -0,0 +1,118 @@
+
+Party::Party(Player* leader, Player* player)
+{
+    AddMember(leader);
+    AddMember(player);
+}
+
+void Party::SendPartyInfo() const
+{
+    std::lock_guard<std::recursive_mutex> guard(m_rmtx_list);
+    packet p(S2C_PARTYINFO);
+    p.push<unsigned char>(GetSize());
+    for(auto& player : m_members_list)
+    {
+        p.push<unsigned int>(player->GetID());
+        p.push_str(player->GetName());
+        p.push<unsigned char>(player->GetClass());
+        p.push<unsigned char>(player->GetLevel());
+        p.push<unsigned short>(player->GetCurHP());
+        p.push<unsigned short>(player->GetMaxHP());
+    }
+    WriteToAll(p);
+}
+
+void Party::WriteToAll(const packet& p) const
+{
+    std::lock_guard<std::recursive_mutex> guard(m_rmtx_list);
+    for(auto& player : m_members_list)
+        player->write(p);
+}
+
+void Party::AddMember(Player* player)
+{
+    std::lock_guard<std::recursive_mutex> guard(m_rmtx_list);
+    player->SetPartyInviterID(0);
+
+    if(IsFull())
+    {
+        player->write(S2C_MESSAGE, "b", MSG_PARTYISFULL);
+        return;
+    }
+
+    m_members_list.push_back(player);
+    player->SetPartyPointer(this);
+
+    if(IsValid())
+    {
+        WriteToAll(packet(S2C_MESSAGE,"b", MSG_JOINEDINPARTY)); //BUG?: msg not working
+        SendPartyInfo();
+    }
+}
+
+void Party::RemoveMember(Player* player, bool isKicked)
+{
+    std::lock_guard<std::recursive_mutex> guard(m_rmtx_list);
+    if(std::find(m_members_list.begin(), m_members_list.end(), player) == m_members_list.end())
+        return;
+
+    bool is_leaver_party_leader = player->IsPartyLeader();
+
+    m_members_list.remove(player);
+    if(isKicked)
+    {
+        player->write(S2C_MESSAGE, "b", MSG_EXILEDFROMPARTY);
+        //TODO: WriteToAll with message that member X was expelled.
+    }
+    else
+    {
+        player->write(S2C_MESSAGE, "b", MSG_LEFTPARTY);
+        //TODO: WriteToAll with message that member X left.
+    }
+
+    if(GetSize() == 1)
```

What if size is equal to 0? If it cannot happen, we should throw an exception

---

done